token-lending/program/src/state/reserve.rs

```diff
@@ -92,6 +92,73 @@ impl Reserve {
         }
     }
 
+    // @TODO: document
+    /// Check obligation health
+    pub fn check_obligation_health(
+        &self,
+        obligation: &Obligation,
+        liquidity_token_mint: &Pubkey,
+        token_converter: impl TokenConverter,
+    ) -> Result<ObligationHealthResult, ProgramError> {
+        Self::_check_obligation_health(
+            obligation,
+            liquidity_token_mint,
+            self.collateral_exchange_rate()?,
+            &self.config,
+            token_converter,
+        )
+    }
+
+    fn _check_obligation_health(
+        obligation: &Obligation,
+        liquidity_token_mint: &Pubkey,
+        collateral_exchange_rate: CollateralExchangeRate,
+        collateral_reserve_config: &ReserveConfig,
+        mut token_converter: impl TokenConverter,
+    ) -> Result<ObligationHealthResult, ProgramError> {
+        // @TODO: does this approach make sense?
+        let borrow_token_price = token_converter.best_price(liquidity_token_mint)?;
+        let decimal_collateral = Decimal::from(obligation.deposited_collateral_tokens);
+        let collateral_value = collateral_exchange_rate
+            .decimal_collateral_to_liquidity(decimal_collateral)?
+            .try_div(borrow_token_price)?;
+        let loan_to_value = obligation
+            .borrowed_liquidity_wads
+            .try_div(collateral_value)?;
+
+        let deposit_amount: u64;
+        let withdraw_amount: u64;
+
+        let reserve_loan_to_value =
+            Decimal::from_percent(collateral_reserve_config.loan_to_value_ratio);
+        // @TODO: does this math make sense?
+        // ratio of the obligation loan to value to reserve loan to value
+        let loan_to_value_ratio = loan_to_value.try_div(reserve_loan_to_value)?;
+        // @TODO: is this comparison safe and idiomatic?
+        if loan_to_value_ratio.lt(&Decimal::one()) {
+            // healthy, up to `withdraw_amount` of collateral could be withdrawn
+            deposit_amount = 0;
+            withdraw_amount = decimal_collateral
+                .try_mul(loan_to_value_ratio)?
+                .try_floor_u64()?;
+        } else if loan_to_value_ratio.gt(&Decimal::one()) {
+            // unhealthy, at least `deposit_amount` of collateral should be deposited
+            deposit_amount = decimal_collateral
+                .try_mul(loan_to_value_ratio)?
+                .try_ceil_u64()?;
+            withdraw_amount = 0;
+        } else {
+            deposit_amount = 0;
+            withdraw_amount = 0;
+        }
```

Comment:
@jstarry My understanding of the ratios:

If an obligation's loan to value ratio is greater than the reserve's loan to value ratio, that implies the loan is getting unhealthy, and additional collateral should be deposited before it gets to the liquidation threshold. Conversely, if an obligation's LTV is less than the reserve LTV, that implies the loan is healthy and some collateral could be withdrawn. The max collateral that should be allowed to be withdrawn is the amount that would bring the obligation's LTV to the reserve LTV.

Therefore, the health of an obligation can be thought of as the ratio of the obligation's LTV to the reserve LTV, which simplifies the math. Is that correct?

Reply:
I asked for some feedback from @joncinque and simplified this implementation in 447c49862b0c33ed04a8559ba2777d75a978820d.